src/shared/firewall-util.c

```diff
@@ -7,21 +7,44 @@
 #include "alloc-util.h"
 #include "firewall-util.h"
 #include "firewall-util-private.h"
+#include "log.h"
+#include "string-table.h"
 
-static enum FirewallBackend firewall_backend_probe(FirewallContext *ctx) {
-        if (fw_nftables_init(ctx) == 0)
-               return FW_BACKEND_NFTABLES;
+static const char * const firewall_backend_table[_FW_BACKEND_MAX] = {
+        [FW_BACKEND_NONE] = "none",
 #if HAVE_LIBIPTC
-        return FW_BACKEND_IPTABLES;
+        [FW_BACKEND_IPTABLES] = "iptables",
+#endif
+        [FW_BACKEND_NFTABLES] = "nftables",
+};
+
+DEFINE_STRING_TABLE_LOOKUP_TO_STRING(firewall_backend, FirewallBackend);
+
+static void firewall_backend_probe(FirewallContext *ctx) {
+        assert(ctx);
+
+        if (ctx->backend != _FW_BACKEND_INVALID)
+                return;
+
+        if (fw_nftables_init(ctx) >= 0)
+                ctx->backend = FW_BACKEND_NFTABLES;
+        else
+#if HAVE_LIBIPTC
+                ctx->backend = FW_BACKEND_IPTABLES;
 #else
-        return FW_BACKEND_NONE;
+                ctx->backend = FW_BACKEND_NONE;
```

Comment:
This approach is OK, but I wonder if it wouldn't make sense to simplify this by skipping the #ifdeffery. Something like this:
```C
if (fw_nftables_init(ctx) >= 0)
   ctx->backend = FW_BACKEND_NFTABLES;
else if (HAVE_LIBIPTC)
     ctx->backend = FW_BACKEND_IPTABLES;
else
    ctx->backend = FW_BACKEND_NONE;
```